library/psa_crypto_storage.h

```diff
@@ -200,8 +188,169 @@ psa_status_t psa_parse_key_data_from_storage( const uint8_t *storage_data,
                                               size_t *key_data_length,
-                                              psa_key_type_t *type,
-                                              psa_key_policy_t *policy );
+                                              psa_key_attributes_t *attributes );
+
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+/** This symbol is defined if transaction support is required. */
+#define PSA_CRYPTO_STORAGE_HAS_TRANSACTIONS
+#endif
+
+#if defined(PSA_CRYPTO_STORAGE_HAS_TRANSACTIONS)
+
+/** The type of transaction that is in progress.
+ */
+/* This is an integer type rather than an enum for two reasons: to support
+ * unknown values when loading a transaction file, and to ensure that the
+ * type has a known size.
+ */
+typedef uint16_t psa_crypto_transaction_type_t;
+
+/** No transaction is in progress.
+ */
+#define PSA_CRYPTO_TRANSACTION_NONE             ( (psa_crypto_transaction_type_t) 0x0000 )
+
+/** A key creation transaction.
+ *
+ * This is only used for keys in an external cryptoprocessor (secure element).
+ * Keys in RAM or in internal storage are created atomically in storage
+ * (simple file creation), so they do not need a transaction mechanism.
+ */
+#define PSA_CRYPTO_TRANSACTION_CREATE_KEY       ( (psa_crypto_transaction_type_t) 0x0001 )
+
+/** A key destruction transaction.
+ *
+ * This is only used for keys in an external cryptoprocessor (secure element).
+ * Keys in RAM or in internal storage are destroyed atomically in storage
+ * (simple file deletion), so they do not need a transaction mechanism.
+ */
+#define PSA_CRYPTO_TRANSACTION_DESTROY_KEY      ( (psa_crypto_transaction_type_t) 0x0002 )
+
+/** Transaction data.
+ *
+ * This type is designed to be serialized by writing the memory representation
+ * and reading it back on the same device.
+ *
+ * \note The transaction mechanism is designed for a single active transaction
+ *       at a time. The transaction object is #psa_crypto_transaction.
+ *
+ * \note If an API call starts a transaction, it must complete this transaction
+ *       before returning to the application.
+ *
+ * The lifetime of a transaction is the following (note that only one
+ * transaction may be active at a time):
+ *
+ * -# Call psa_crypto_prepare_transaction() to initialize the transaction
+ *    object in memory and declare the type of transaction that is starting.
+ * -# Fill in the type-specific fields of #psa_crypto_transaction.
+ * -# Call psa_crypto_save_transaction() to start the transaction. This
+ *    saves the transaction data to internal storage.
+ * -# If there are intermediate stages in the transaction, update
+ *    the fields of #psa_crypto_transaction and call
+ *    psa_crypto_save_transaction() again when each stage is reached.
+ * -# When the transaction is over, whether it has been committed or aborted,
+ *    call psa_crypto_stop_transaction() to remove the transaction data in
+ *    storage and in memory.
```

Comment:
I think you have a specific transaction model in mind that works by preparing the new data without touching the old data and, as the last step, atomically switching to the new data (“commit”). But that's not how transactions work here. There's no way to atomically switch to the new data. Any distinction between “abort” and “commit” is in the eyes of the caller. As far as the transaction storage subsystem is concerned, there's just stopping.

Reply:
How does anything happen? I see transaction put into storage, nobody using it, and then removed. This has no effect. How do we expect this to work? It may be that we make it work in a follow up PR, but I am curious how we'd want this to work.

I'd expect core to read the transaction (if any) and commit/apply it atomically with other data it is storing, but I don't see any signs of that intent.

Reply:
We probably need a bit more commentary on the transaction model in use.

Because the underlying storage provides single-item atomic replacement, it would be possible to build a multi-item atomic update on top with a suitably designed heirarchy of stored objects. But this measn the the core needs to maintain a directory of data uids which it replaces when running a transaction. Although this is more flexible and permits complex changes to be built into a transaction - we don't actually need this facility.

The model Gilles has built is more like a single-item roll-forward journal, or perhaps even better described as a persistent state machine. The transaction object is just the journal entry, and it records all the data that is needed for the distributed update required to either finish the key operation or roll it back. It also is not generic - it stores operation specific data in order for operation-specific completion/rollback code to handle the situation where the transaction is found on restart.

So for each operation that requires multiple operations on persistent storage, we define a state machine where the current state is stored in the "transaction object". The stored transaction state is enough to permit either a "roll-forward" or a "roll-back" to the key store state - and provide the atomicity required for key store changes.  The recovery mode (rollback/rollfoward), can change for a operation depending on the state-machine state, e.g. when creating a key it will be roll-back until we know the SE has created/stored the key material successfully, the transaction object after that can record a state that allows roll-forward as well as roll-back.

Note that rolling back might fail in some circumstances, as this can require one or more changes to persistent data to accomplish. If it fails, the system could operate by acting "as if the rollback had completed" and retaining the transaction object so that this is remembered. However, that will block any key store operation which would use a new transaction.

Maybe some good verbs for the API would be 'start', 'update' and 'finish', (or 'begin' and 'end') with regards to operations on the persistent transaction state. 'commit', 'abort' and 'stop' may fit a classic transaction model (thinking databases here), but each key store operation that uses this transaction model has a different 'commit point' - when the stored transaction state changes from "must roll back" to "can roll forward"

Reply:
@Patater The storage subsystem provides primitives to store the transaction state. It does not perform the transaction.

The core only ever needs to read back the transaction data if the transaction is interrupted by a reset. (At least so far, this is not intrinsic in the design.)